python/tests/server/test_worker.py

```diff
@@ -465,42 +498,57 @@ def __init__(self):
         self.predict_generator = None
         self.predict_events = []
         self.predict_payload = None
+        self.setup_done = False
+
+    def await_(self, coro: Awaitable[T]) -> T:
+        return self.loop.run_until_complete(coro)
 
     @rule(sleep=st.floats(min_value=0, max_value=0.5))
     def wait(self, sleep):
         time.sleep(sleep)
 
+    @precondition(lambda x: not x.setup_done)
     @rule()
     def setup(self):
+        trace("did setup")
         try:
             self.setup_generator = self.worker.setup()
             self.setup_events = []
+        # this except allows setup to be called many times which seems weird
         except InvalidStateException:
             pass
 
-    @precondition(lambda x: x.setup_generator)
-    @rule(n=st.integers(min_value=1, max_value=10))
-    def read_setup_events(self, n):
+        # read_setup_events being a separate rule seems like it was already wrong
+        # hypothesis would call setup repeatedly, just now the InvalidStateException
+        # comes up when you actually use the iterator
+
+        # @precondition(lambda x: x.setup_generator)
+        # @rule(n=st.integers(min_value=1, max_value=10))
+        # def read_setup_events(self, n):
         try:
-            for _ in range(n):
-                event = next(self.setup_generator)
+            while 1:  # we want all the events, not just "n" events
+                event = self.await_(anext(self.setup_generator))
                 self.setup_events.append(event)
-        except StopIteration:
+        # if we wanted to keep the InvalidStateException to ignore setup assert_state(NEW)
+        # we would put it here, where the exception is raised now
+        except (StopIteration, StopAsyncIteration):
             self.setup_generator = None
+            self.await_(self._check_setup_events())
 
-            self._check_setup_events()
-
-    def _check_setup_events(self):
+    async def _check_setup_events(self):
         assert isinstance(self.setup_events[-1], Done)
 
-        print(self.setup_events)
-        result = _process(self.setup_events)
+        trace(self.setup_events)
+        result = await _process(self.setup_events)
         assert result.stdout == "did setup\n"
         assert result.stderr == ""
         assert result.done == Done()
+        self.setup_done = True
 
+    @precondition(lambda x: x.setup_done)
```

Comment:
Again, this isn't right. By adding this precondition you're removing branches from the tree. Creating a predictor and calling `predict` before `setup` is a thing that might happen. Without this precondition we're checking that the thing behaves "correctly" if that occurs -- i.e. it throws `InvalidStateException`.